### PR TITLE
Cleanup `hostdevice_vector` and add more APIs

### PR DESCRIPTION
**Pull request by ttnghia**
This work includes:
 * Fix a bug in `hostdevice_vector` when the host buffer does not change its size when appending new elements. Instead, the new elements are written directly into raw memory (which is out of bounds). Previously, this did not trigger any issue since the host buffer has reserved plenty of memory upon its construction, until I attempted to access the `front()` and `back()` elements of it.
 * Add `front()` and `back()` accessors which return the first and last elements in the host buffer.

### CODE REVIEW

**Comment by vuule** [comment]
how did this come up?
I just talked to @nvdbaranec about removing `push_back` from `hostdevice_vector`, but it sounds like you are actively using that feature.

**Comment by ttnghia** [comment]
I've just merged 15079 into my chunked ORC reader and all the tests crashed, so I figured out this bug.

**Comment by ttnghia** [comment]
> how did this come up? I just talked to @nvdbaranec about removing `push_back` from `hostdevice_vector`, but it sounds like you are actively using that feature.

Indeed I'm not using `push_back`. I'm using `front` and `back()` in my ORC chunked reader. So, removing `push_back` would be fine.

**Comment by ttnghia** [comment]
I've just tried to remove `push_back` and found that only one place in the ORC reader is using it. So I think we can remove it quickly in this PR too. Please let me know if you want that.

**Comment by vuule** [comment]
> I've just tried to remove `push_back` and found that only one place in the ORC reader is using it. So I think we can remove it quickly in this PR too. Please let me know if you want that.
> 
> Note that this will change the interface of `hostdevice_vector`, removing the `initial_size` parameter completely. From now, we will always construct host/device vector at maximum size.

If you think you can quickly remove it, please go ahead. I'll be happy to see `initial_size` and `current_size` go away.

**Comment by bdice** [comment]
No push_back… not resizable… hmm. Hey, that’s not a vector at all!

image

That’s an array!

**Comment by ttnghia** [comment]
`hostdevice_array` LGTM :)

**Comment by vuule** [comment]
@bdice first of all, :joy:
I guess it is more of a buffer/array (array to me implies compile-time size). Too bad thrust's buffers are untyped, `hostdevice_buffer` sounds great to me.

File: cpp/src/io/orc/reader_impl_preprocess.cu [hunk 1]
```diff
@@ -180,12 +180,11 @@ rmm::device_buffer decompress_stripe_data(
...
-    0, stream_info.size(), stream);
-  for (auto const& info : stream_info) {
-    compinfo.push_back(gpu::CompressedStreamInfo(
-      static_cast<uint8_t const*>(stripe_data[info.stripe_idx].data()) + info.dst_pos,
-      info.length));
+  cudf::detail::hostdevice_vector<gpu::CompressedStreamInfo> compinfo(stream_info.size(), stream);
+  for (std::size_t idx = 0; idx < stream_info.size(); ++idx) {
+    auto const& info = stream_info[idx];
+    compinfo[idx]    = gpu::CompressedStreamInfo(
+      static_cast<uint8_t const*>(stripe_data[info.stripe_idx].data()) + info.dst_pos, info.length);
```
Comment (vuule) [inline cpp/src/io/orc/reader_impl_preprocess.cu:187]: Are you kidding me, this was it?!

**Comment by vuule** [comment]
apparently span tests also use push_back

**Comment by ttnghia** [comment]
So far we have the new names:
 * `hostdevice_vector` (current)
 * `hostdevice_array`
 * `hostdevice_buffer`

Any other candidate? @nvdbaranec?

File: cpp/src/io/orc/reader_impl_preprocess.cu (hunk 1 above)
Comment (vuule) [reply cpp/src/io/orc/reader_impl_preprocess.cu:187]: Nothing, I just thought (feared) bigger changes were needed to remove push_back.

**Comment by vuule** [comment]
> So far we have the new names:
> 
> * `hostdevice_vector` (current)
> * `hostdevice_array`
> * `hostdevice_buffer`
> 
> Any other candidate? @nvdbaranec?

Not sure if this is clear - IMO we should not rename it in this PR. Maybe something we can evaluate for 24.06.

**Comment by nvdbaranec** [comment]
I'd point out: `rmm::device_uvector` doesn't support push_back either.  It does support `resize` though.

**Comment by nvdbaranec** [comment]
> @bdice first of all, 😂 I guess it is more of a buffer/array (array to me implies compile-time size). Too bad thrust's buffers are untyped, `hostdevice_buffer` sounds great to me.

For me "buffer" implies typeless, ie `rmm::device_buffer`.  Just bytes.

**Comment by vuule** [comment]
we could add a `resize` if that helps the argument for not renaming the class :shrug:

**Comment by bdice** [comment]
I don’t want to force a name change in this PR. Just had to ~make a meme~ share the idea. Let’s discuss in an issue, as it seems this PR isn’t 100% sure on the API yet.

**Comment by ttnghia** [comment]
I think we should better keep the `push_back` API as it is more convenient to use. My upcoming chunked ORC reader has more of its usage. Without it, we have to create an index variable and keep track of it to write elements iteratively.

So, keep `push_back`, and keep the name `hostdevice_vector` unchanged?

**Comment by vuule** [comment]
> So, keep `push_back`, and keep the name `hostdevice_vector` unchanged?

Feel free to limit this PR to its original scope.

I always felt like `push_back` was half-baked in this class and thus should be removed. Maybe instead we need to make sure it's fully baked. Either way, nothing urgent for this PR/release.

**Review by nvdbaranec — APPROVED** [review]

**Review by vuule — APPROVED** [review]

**Pull request merged**